File: sorting/select-sort/select-sort.js
// 交换函数
function swap(array, i, j) {
  const temp = array[i];
  array[i] = array[j];
  array[j] = temp;
}

// 选择排序
function selectSort(array) {
  let indexOfMin;           // 用于标识每趟比较中最小值的索引
  const size = array.length;
  
  for (let i = 0; i < size - 1; i++) {
    indexOfMin = i;         // 每次初始化最小值的索引为第一个
    
    for (let j = i + 1; j < size; j++) {
      if (array[j] < array[indexOfMin]) {
        indexOfMin = j;     // 如果找到更小的值则更新indexOfMin
      }
    }
    
    // 交换
    if (i !== indexOfMin) {
      swap(array, i, indexOfMin);
    }
  }
}

let array = [39, 28, 57, 12, 95, 45, 10, 73];

selectSort(array);

console.log(array);

File: sorting/select-sort/SelectSort.java
package algorithms;

public class Sorting {
    // 交换方法
    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    // 选择排序
    public static void selectSort(int[] array) {
        int i, j, indexOfMin;
        int size = array.length;

        for (i = 0; i < size - 1; i++) {
            indexOfMin = i;

            for (j = i + 1; j < size; j++) {
                if (array[j] < array[indexOfMin]) {
                    indexOfMin = j;
                }
            }

            if (i != indexOfMin) {
                swap(array, i, indexOfMin);
            }
        }
    }

    public static void main(String[] args) {
        int[] array = {39, 28, 57, 12, 95, 45, 10, 73};

        Sorting.selectSort(array);

        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }
    }
}
